### ISSUE TITLE
[Need advice] `ExecutorRef::default()` is considered as `Optional<UnownedSerialExecutor>(nil)`

### ISSUE THREAD

**Issue by kateinoigakukun** [apple/swift#65529]
**Description**

`ExecutorRef::default()` is used as the main executor in cooperative global executor mode.
So `MainActor.sharedUnownedExecutor` returns `ExecutorRef::default` (`{i64 0, i64 0}`) in the mode.

However, the `Builtin.Executor` wrapped by `UnownedSerialExecutor` has [some extra inhabitants including zero representation](https://github.com/apple/swift/blob/6ae1077720402e71d2fba41c6996f0d93daacad7/lib/IRGen/GenConcurrency.cpp#L77-L110), so `{i64 0, i64 0}` is considered as nil when interpreting it as `Optional<UnownedSerialExecutor>` (or `Optinal<Builtin.Executor>`).
According to https://github.com/apple/swift/pull/36878/commits/ec5215bf4f0a75e0087d2066fefd2c4bd8c33309, the builtin type preserved 0 before but removed the preservation intentionally for some reasons (for simplicity?).

Therefore, `Optional.some(MainActor.sharedUnownedExecutor)` is `nil` in cooperative global executor mode now.
(I found this when checking a failure of [distributed_actor_custom_executor_from_id.swift](https://github.com/apple/swift/blob/6ae1077720402e71d2fba41c6996f0d93daacad7/test/Distributed/Runtime/distributed_actor_custom_executor_from_id.swift#L45-L47) in [WebAssembly CI](https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-20.04-webassembly/1228/console)

So my question here is:
1. Should we avoid exposing `ExecutorRef::default()` to Swift world?
2.a. If 1 is no, does it make sense to restore the implicit inhabitants in Builtin.Executor?
2.b. If 1 is yes, does it make sense to have a separate main executor and expose it to Swift instead of exposing `default()`.

**Potential related issue**

The identity field of `ExecutorRef` can be [`DISPATCH_QUEUE_GLOBAL_EXECUTOR`, which is defined as `1`](https://github.com/apple/swift/blob/6ae1077720402e71d2fba41c6996f0d93daacad7/stdlib/public/Concurrency/TaskPrivate.h#L114). Fortunately, there is no interface to get a global executor in Swift world, but we have to preserve 1 also in theory if I understand correctly.

It would be very appreciated if I could get some advice from concurrency people (especially @rjmccall)

**Comment by ktoso** [comment]
I think you mean `ExecutorRef::generic()`, we don't spell it as "default". But yes it is (0, 0) and has this issue.

I don't know the exact thinking @rjmccall had with that patch, but I can give some thoughts in general?

We don't offer APIs today to get a reference to the default / global / generic pool in Swift, so until we do we should probably avoid exposing anything like it in Swift.

If I'm seeing this correctly DISPATCH_QUEUE_GLOBAL_EXECUTOR seems to only be used from inside Dispatch executor when we "enqueue on global" and there's a special handling to return ExecutorRef::generic() for it somewhere...

The comment about it being main or default executor seems outdated tbh? It came from code like this:
```
/// Special values used with DispatchQueueIndex to indicate the global and main
/// executors.
#define DISPATCH_QUEUE_GLOBAL_EXECUTOR (void *)1
#define DISPATCH_QUEUE_MAIN_EXECUTOR (void *)2 // not true
```

which isn't true anymore -- the main executor isn't using 2, but the exact "main queue", so the entire comment makes me feel like it may not be quite right.

Not sure if this helps @kateinoigakukun ?

Would be good to see what @rjmccall has to say about the nil behavior since he made that change a while back

**Comment by rjmccall** [comment]
I think that implementation is probably just wrong.  We ought to find some way to distinguish the main executor from the `nil` executor.

I don't remember much about the "cooperative executor" mode, even though I suspect I wrote it.  Is this really being used by someone?  Is the implicit expectation that there's only one thread?

**Comment by ktoso** [comment]
Thank you for chiming in John! I read too much meaning into that patch then I guess. 

Cooperative executor mode I think is the flag forcing things to be one thread and wasm uses that which is why @kateinoigakukun is asking I think? If I understand correctly.

**Comment by kateinoigakukun** [comment]
Yes, cooperative executor is used in Wasm where we can't assume multi-threads.

> We ought to find some way to distinguish the main executor from the nil executor, even if they end up adding jobs to the same queue behind the scenes.

Ok, thank you for the clarification! I will make time to fix this later.
